idx/memory/meta_tags.go

```diff
@@ -0,0 +1,236 @@
+package memory
+
+import (
+	"fmt"
+	"hash"
+	"hash/fnv"
+	"sort"
+	"strings"
+)
+
+// the collision avoidance window defines how many times we try to find a higher
+// slot that's free if two record hashes collide
+const collisionAvoidanceWindow = 3
+
+// the function we use to get the hash for hashing the meta records
+// it can be replaced for mocking in tests
+var queryHash func() hash.Hash32
+
+func init() {
+	queryHash = fnv.New32a
+}
+
+type metaTagRecord struct {
+	metaTags []kv
+	queries  []expression
+}
+
+// list of meta records keyed by a unique identifier used as ID
+type metaTagRecords map[uint32]metaTagRecord
+
+// index structure keyed by tag -> value -> list of meta record IDs
+type metaTagValue map[string][]uint32
+type metaTagIndex map[string]metaTagValue
+
+func (m metaTagIndex) deleteRecord(keyValue kv, recordId uint32) {
+	if values, ok := m[keyValue.key]; ok {
+		if recordIds, ok := values[keyValue.value]; ok {
+			for i := 0; i < len(recordIds); i++ {
+				if recordIds[i] == recordId {
+					// no need to keep the order
+					recordIds[i] = recordIds[len(recordIds)-1]
+					values[keyValue.value] = recordIds[:len(recordIds)-1]
+
+					// no id should ever be present more than once
+					return
+				}
+			}
+		}
+	}
+}
+
+func (m metaTagIndex) insertRecord(keyValue kv, recordId uint32) {
+	var values metaTagValue
+	var ok bool
+
+	if values, ok = m[keyValue.key]; !ok {
+		values = make(metaTagValue)
+		m[keyValue.key] = values
+	}
+
+	values[keyValue.value] = append(values[keyValue.value], recordId)
+}
+
+// metaTagRecordFromStrings takes two slices of strings, parses them and returns a metaTagRecord
+// The first slice of strings has the meta tags & values
+// The second slice has the tag query expressions which the meta tags & values refer to
+// On parsing error the second returned value is an error, otherwise it is nil
+func metaTagRecordFromStrings(metaTags []string, tagQueryExpressions []string) (metaTagRecord, error) {
```

Comment:
The suggestion to move the parsing/validation to an earlier stage makes sense. At the moment this is here for consistency with all the other tag query methods, such as: https://github.com/grafana/metrictank/blob/947555d9856ab3c0da297e0be8be2fb163786c90/idx/memory/memory.go#L1012

It would probably make sense to move the whole expression parsing/validation for all the tag query handling methods to an earlier stage in the request handling. But I wouldn't want to do that in this PR, otherwise this PR will also need to modify a lot of the existing tag query call-paths.